common/core/desktop/src/kmx/kmx_xstring.cpp

```diff
@@ -119,33 +119,57 @@ km_kbp_cp *km::kbp::kmx::u16tok(km_kbp_cp *p, km_kbp_cp ch, km_kbp_cp **ctx) {
 
 PKMX_WCHAR km::kbp::kmx::incxstr(PKMX_WCHAR p)
 {
-  if(*p == 0) return p;
-  if(*p != UC_SENTINEL)
+  int deltaptr;			// how many bytes to jump over 
+
+  if (*p == 0) return p;
+  if (*p != UC_SENTINEL)
   {
-    if(*p >= 0xD800 && *p <= 0xDBFF && *(p+1) >= 0xDC00 && *(p+1) <= 0xDFFF) return p+2;
-    return p+1;
+    if (*p >= 0xD800 && *p <= 0xDBFF && *(p + 1) >= 0xDC00 && *(p + 1) <= 0xDFFF) return p + 2;
+    return p + 1;
   }
-
-  p+=2;
-  switch(*(p-1))
+  else
   {
-    case CODE_ANY:      return p+1;
-    case CODE_NOTANY:   return p+1;
-    case CODE_INDEX:    return p+2;
-    case CODE_USE:      return p+1;
-    case CODE_DEADKEY:    return p+1;
-    case CODE_EXTENDED:   p += 2; while(*p && *p != UC_SENTINEL_EXTENDEDEND) p++; return p+1;
-    case CODE_CLEARCONTEXT: return p+1;
-    case CODE_CALL:     return p+1;
-    case CODE_CONTEXTEX:  return p+1;
-    case CODE_IFOPT:    return p+3;
-    case CODE_IFSYSTEMSTORE: return p+3;
-    case CODE_SETOPT:   return p+2;
-    case CODE_SETSYSTEMSTORE: return p+2;
-    case CODE_RESETOPT: return p+1;
-    case CODE_SAVEOPT:  return p+1;
-    default:        return p;
+    // UC_SENTINEL(FFFF) with UC_SENTINEL_EXTENDEDEND(0x10) == variable length
+    if (*(p + 1) == CODE_EXTENDED) {
+      p += 2;
+      while (*(p - 1) && *p && *p != UC_SENTINEL_EXTENDEDEND)
```

Comment:
```suggestion
      while (*p && *p != UC_SENTINEL_EXTENDEDEND)
```

We've just tested p-1 (was p+1) on line 133. We don't need to test it again.

Reply:
OK